HPCC-25212 Add support for unity builds in CMake

jpmcmu opened the pull request:
WIP to reduce compile time of the platform. At the moment working on testing impact of enabling unity builds.

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [X] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [X] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [X] The commit message is properly formatted and free of typos.
  - [X] The commit message title makes sense in a changelog, by itself.
  - [X] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [X] I have read the CONTRIBUTORS document.
- [X] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [X] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

HPCCSmoketest commented:
Process of PR-14292, label: UNITY-BUILD starts now.
The reason of this test is: Forced to re-test.
Commit ID: 3e8d813d737d2eb0640da8cd9f3151930af3b879
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-148.ca-central-1.compute.internal

jpmcmu commented:
Creating this to share some experiments to reduce compile time of the platform. Currently have enabled unity builds on some of the sub-projects. Enabled unity build on some of the projects that don't require as much work to fix duplicate symbols etc. Enabling on all projects would further cut down compile time another 4-5 minutes.

Precompiled headers for things like jlib would also probably have a decent impact on compile time.

Timing of make -j6 on my VM:

Without Unity Builds:
real    31m1.788s
user    115m24.239s
sys     6m0.177s

Unity Builds:
real    21m47.372s
user    80m47.214s
sys     4m0.005s

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-148.ca-central-1.compute.internal
Sha: 3e8d813d737d2eb0640da8cd9f3151930af3b879
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |0 |11 |5 sec  (00:00:05) |
|setup (thor) |11 |0 |11 |5 sec  (00:00:05) |
|setup (roxie) |11 |0 |11 |5 sec  (00:00:05) |
|test (hthor) |0 |0 |0 |N/A |
|test (thor) |0 |0 |0 |N/A |
|test (roxie) |0 |0 |0 |N/A |


[Errors:](https://gist.github.com/4ab5f3406f1e94552eb4f50173de0ac1)
- setup
  - hthor
    - 1\. [roxie_echo](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-roxie_echo-no-wuid-log), 2\. [roxie_keepwhitespace](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-roxie_keepwhitespace-no-wuid-log), 3\. [setup](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setup-w20201111-160526-1-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setup_fetch-w20201111-160526-3-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupdict-w20201111-160526-9-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setuppersist-w20201111-160526-8-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupsearchindex-w20201111-160526-4-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupsq-w20201111-160526-6-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupwordindex-w20201111-160526-5-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupxml-w20201111-160526-2-log), 11\. [soapbase](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-soapbase-no-wuid-log), 
  - thor
    - 1\. [roxie_echo](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-roxie_echo-no-wuid-log), 2\. [roxie_keepwhitespace](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-roxie_keepwhitespace-no-wuid-log), 3\. [setup](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setup-w20201111-160531-log), 4\. [setup_fetch](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setup_fetch-w20201111-160531-1-log), 5\. [setupdict](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupdict-w20201111-160531-3-log), 6\. [setuppersist](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setuppersist-w20201111-160531-8-log), 7\. [setupsearchindex](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupsearchindex-w20201111-160531-6-log), 8\. [setupsq](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupsq-w20201111-160531-5-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupwordindex-w20201111-160531-2-log), 10\. [setupxml](https://gist.github.com/HPCCSmoketest/4ab5f3406f1e94552eb4f50173de0ac1#file-setupxml-w20201111-160531-1
 ... ( comment is too long, 48 lines, 5696 bytes, truncated) ... 
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  600 sec (00:10:00) |  106 sec (00:01:46) |   20 sec (00:00:20) |   16 sec (00:00:16) |  222 sec (00:03:42) |   14 sec (00:00:14) |  988 sec (00:16:28) |

```

HPCCSmoketest commented:
Process of PR-14292, label: UNITY-BUILD starts now.
The reason of this test is: Forced to re-test.
Commit ID: 59d45ae8c2216d3e224579ecbb88f743d4649605
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-138.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-138.ca-central-1.compute.internal
Sha: 59d45ae8c2216d3e224579ecbb88f743d4649605
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |51 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |0 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |6 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |11 |11 |0 |24 sec  (00:00:24) |
|setup (thor) |11 |11 |0 |40 sec  (00:00:40) |
|setup (roxie) |11 |11 |0 |22 sec  (00:00:22) |
|test (hthor) |936 |936 |0 |316 sec  (00:05:16) |
|test (thor) |847 |847 |0 |655 sec  (00:10:55) |
|test (roxie) |1015 |1015 |0 |352 sec  (00:05:52) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  591 sec (00:09:51) |  104 sec (00:01:44) |   20 sec (00:00:20) |   16 sec (00:00:16) | 1632 sec (00:27:12) |   17 sec (00:00:17) | 2389 sec (00:39:49) |

jpmcmu commented:
@rpastrana please review

jpmcmu commented:
@richardkchapman @ghalliday 

Ran some comparisons between ninja and make. Ninja didn't really provide any improvement on my VM. It could be that  Ninja would provide a benefit with more cores? (Disabled AWS, Azure, WS_SQL to get ninja to work, fixes to those projects wouldn't be too difficult if we want to get ninja working for them)

Make without Unity Build
real    20m1.892s
user    74m3.783s
sys     3m44.617s

Ninja without Unity Build
real    19m43.717s
user    74m49.448s
sys     3m34.738s

----------------------------

Ninja with Unity Build
real    15m39.071s
user    59m46.666s
sys     2m31.265s

Make with Unity Build
real    15m36.859s
user    57m37.442s
sys     2m33.697s

ghalliday commented:
Thanks, James.  What is your system configuration?
Gavin

________________________________
From: James McMullan <notifications@github.com>
Sent: 13 November 2020 21:58
To: hpcc-systems/HPCC-Platform <HPCC-Platform@noreply.github.com>
Cc: Halliday, Gavin (RIS-HBE) <Gavin.Halliday@lexisnexisrisk.com>; Mention <mention@noreply.github.com>
Subject: Re: [hpcc-systems/HPCC-Platform] WIP: Compile time reduction (#14292)


*** External email: use caution ***



@richardkchapman<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Frichardkchapman&data=04%7C01%7CGavin.Halliday%40lexisnexisrisk.com%7C409a5cf808094914d68c08d8881f4af8%7C9274ee3f94254109a27f9fb15c10675d%7C0%7C0%7C637409015272930371%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=MqkwpKqWV%2FLOmyo2NsdZrDdXdZEES9DEsYEtFlxIZoQ%3D&reserved=0> @ghalliday<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fghalliday&data=04%7C01%7CGavin.Halliday%40lexisnexisrisk.com%7C409a5cf808094914d68c08d8881f4af8%7C9274ee3f94254109a27f9fb15c10675d%7C0%7C0%7C637409015272930371%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=tvghoEcqq7Jz57sUVYNF98mvs%2BkYb0PtW2G0ujIcvQQ%3D&reserved=0>

Ran some comparisons between ninja and make. Ninja didn't really provide any improvement on my VM. It could be that Ninja would provide a benefit with more cores? (Disabled AWS, Azure, WS_SQL to get ninja to work, fixes to those projects wouldn't be too difficult if we want to get ninja working for them)

Make without Unity Build
real 20m1.892s
user 74m3.783s
sys 3m44.617s

Ninja without Unity Build
real 19m43.717s
user 74m49.448s
sys 3m34.738s

________________________________

Ninja with Unity Build
real 15m39.071s
user 59m46.666s
sys 2m31.265s

Make with Unity Build
real 15m36.859s
user 57m37.442s
sys 2m33.697s

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fhpcc-systems%2FHPCC-Platform%2Fpull%2F14292%23issuecomment-727056456&data=04%7C01%7CGavin.Halliday%40lexisnexisrisk.com%7C409a5cf808094914d68c08d8881f4af8%7C9274ee3f94254109a27f9fb15c10675d%7C0%7C0%7C637409015272940321%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=KpdxEuu7Pu0mt917vFoJ2e8rpDLv4Xu7cSNVhOse918%3D&reserved=0>, or unsubscribe<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAAG4KWSXQ2GOL5L5MJU3AKDSPWTZHANCNFSM4TSETW3A&data=04%7C01%7CGavin.Halliday%40lexisnexisrisk.com%7C409a5cf808094914d68c08d8881f4af8%7C9274ee3f94254109a27f9fb15c10675d%7C0%7C0%7C637409015272940321%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=wrXCZyFvMApCsScSjCk2tj0kMOxSCP9ARwFMqXbtkAc%3D&reserved=0>.

________________________________

This email is from RELX Group plc, a company registered in England and Wales with company number 2746616, whose registered office is 1-3 Strand, London, WC2N 5JR, United Kingdom.

The information transmitted is intended only for the person or entity to which it is addressed and may contain confidential and/or privileged material. If you are not the intended recipient of this message, please do not read, copy, use or disclose this communication and notify the sender immediately. It should be noted that any review, retransmission, dissemination or other use of, or taking action in reliance upon, this information by persons or entities other than the intended recipient is prohibited.

________________________________
The information contained in this e-mail message is intended only for the personal and confidential use of the recipient(s) named above. This message may be an attorney-client communication and/or work product and as such is privileged and confidential. If the reader of this message is not the intended recipient or an agent responsible for delivering it to the intended recipient, you are hereby notified that you have received this document in error and that any review, dissemination, distribution, or copying of this message is strictly prohibited. If you have received this communication in error, please notify us immediately by e-mail, and delete the original message.

jpmcmu commented:
@ghalliday The VM I was compiling in has 4 cores and 8Gb

ghalliday commented:
A similar set of timings from a machine with 16 cores, and plenty of memory:

Make without Unity Build
real	6m6.840s
user	70m35.321s
sys	4m29.613s

Ninja without Unity Build
real	4m52.003s
user	71m1.937s
sys	4m19.930s

With unity (with cppunit also disabled)

Make with Unity Build
real	5m47.557s
user	51m57.114s
sys	2m59.358s

Ninja with Unity Build
real	4m12.480s
user	52m40.994s
sys	2m55.314s


Which suggests if you have plenty of cores that ninja has a much bigger effect than unity!

jpmcmu commented:
@ghalliday Ah interesting. Out of curiosity what was the -j option you gave make? I have heard that make works best with a j value of 150% of the core count. Curious to see if that still holds up at 16 cores

ghalliday commented:
That was -j18.  There are a few stages where the cpu dropped significantly below all cores in use. ninja seemed to keep them all active.

jpmcmu commented:
Hmm I wonder if the 150% -j24 would do a better job keeping all cores utilized?

ghalliday commented:
I don't think so...  -j24:

real	6m0.963s
user	68m31.254s
sys	4m19.242s

make -j (overcommitted with no number)

real	6m30.859s
user	69m36.479s
sys	4m24.983s


I think the issue is that some of the projects can't use all the threads, and in those cases make doesn't do a great job of building other things in parallel.

ghalliday commented:
My impression is that unity, ninja and possibly precompiled headers may all have a significant impact on build times.  I think for this PR I would add an option USE_UNITY into commonSetup which can then be used to easily toggle the functionality on and off.

ghalliday commented:
@jpmcmu I hadn't assigned back to you.  I think it is worth pursing, and adding a global option to enable it - so that it is easy to turn on and off.  It will need rebasing once changes are upmerged from 7.12.x to master.

ghalliday commented:
@jpmcmu are you planning on clening up this branch, or would you prefer me to do it?

jpmcmu commented:
@ghalliday I will clean it up and get it ready to go this week; I wouldn't want to add more work to your list. There are still a few projects that don't have unity build support, but it probably makes more sense to add unity support for those projects that in a future PR.

ghalliday commented:
yes, I would gradually add support to other parts.  This should be enough to speed some systems up siginficantly.

jpmcmu committed:
HPCC-25212 Add support for unity builds in CMake

- Added option for enabling unity builds in CMake
- Made minor changes to allow unity builds to compile
- Added unity build support to a subset of projects

Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

HPCCSmoketest commented:
Process of PR-14292, label: UNITY-BUILD starts now.
The reason of this test is: Forced to re-test.
Commit ID: 2a109ad24c7f6fa4283513fe16dcdb22a470aa26
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-146.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :x:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-146.ca-central-1.compute.internal
Sha: 2a109ad24c7f6fa4283513fe16dcdb22a470aa26
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |25 sec  (00:00:25) |
|setup (thor) |9 |9 |0 |42 sec  (00:00:42) |
|setup (roxie) |19 |19 |0 |21 sec  (00:00:21) |
|test (hthor) |947 |947 |0 |634 sec  (00:10:34) |
|test (thor) |858 |858 |0 |1098 sec  (00:18:18) |
|test (roxie) |1020 |1019 |1 |1876 sec  (00:31:16) |


[Errors:](https://gist.github.com/523919344e3f67fafbd871f90f64a11d)
- test
  - roxie
    - 662\. [schedule1](https://gist.github.com/HPCCSmoketest/523919344e3f67fafbd871f90f64a11d#file-schedule1-w20210120-152853-3-log), 

HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    9 sec (00:00:09) |  614 sec (00:10:14) |  124 sec (00:02:04) |   20 sec (00:00:20) |   16 sec (00:00:16) | 2189 sec (00:36:29) |   16 sec (00:00:16) | 2988 sec (00:49:48) |

jpmcmu commented:
@ghalliday please review

HPCCSmoketest commented:
Process of PR-14292, label: UNITY-BUILD starts now.
The reason of this test is: Forced to re-test.
Commit ID: 2a109ad24c7f6fa4283513fe16dcdb22a470aa26
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-123.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-123.ca-central-1.compute.internal
Sha: 2a109ad24c7f6fa4283513fe16dcdb22a470aa26
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |145 |145 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |22 sec  (00:00:22) |
|setup (thor) |9 |9 |0 |41 sec  (00:00:41) |
|setup (roxie) |19 |19 |0 |17 sec  (00:00:17) |
|test (hthor) |947 |947 |0 |667 sec  (00:11:07) |
|test (thor) |858 |858 |0 |1133 sec  (00:18:53) |
|test (roxie) |1020 |1020 |0 |726 sec  (00:12:06) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    8 sec (00:00:08) |  604 sec (00:10:04) |  111 sec (00:01:51) |   23 sec (00:00:23) |   16 sec (00:00:16) | 1436 sec (00:23:56) |   18 sec (00:00:18) | 2216 sec (00:36:56) |

ghalliday commented on ecl/eclcmd/eclcmd_common.cpp:
```diff
@@ -31,6 +31,7 @@ static StringBuffer eclccpath;
 
 const char *queryEclccPath(bool optVerbose)
 {
+    static StringBuffer eclccpath;
```
Should the definition on line 30 have been deleted?

ghalliday commented on system/jlib/jatomic.hpp:
```diff
@@ -197,10 +197,13 @@ typedef volatile long atomic_t;
-typedef struct { volatile int counter; } atomic_t;
 #define ATOMIC_INIT(i)          { (i) }
 #define atomic_read(v)          ((v)->counter)
+
```
This change doesn't match the one in eclrtl_imp.hpp.  That added

#ifndef atomic_type_defined

All of this code should be deleted anyway, but I have created a separate jira for that.

ghalliday reviewed (changes requested):
@jpmcmu It looks good.  The cmake docs suggest it should be implemented slightly differently though:

> Projects should not directly set the UNITY_BUILD property or its associated CMAKE_UNITY_BUILD variable to true. .... However, it IS recommended to set the UNITY_BUILD target property to false if it is known that enabling unity builds for the target can lead to problems.

https://cmake.org/cmake/help/latest/prop_tgt/UNITY_BUILD.html#prop_tgt:UNITY_BUILD

I think this would mean all the places where it is set to true are removed (and a different mechanism is used to disable it?)

jpmcmu committed:
WIP

HPCCSmoketest commented:
Process of PR-14292, label: UNITY-BUILD starts now.
The reason of this test is: Forced to re-test.
Commit ID: 5003a1ef5bfcdba131863687ec06941dc879fdbf
Estimated completion time is ~0.75 hour(s)
centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-100.ca-central-1.compute.internal

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.6.1810 (Linux 3.10.0-957.1.3.el7.x86_64)
 GCC: gcc (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
 Host: ip-10-20-0-100.ca-central-1.compute.internal
Sha: 5003a1ef5bfcdba131863687ec06941dc879fdbf
Build: success
Milestone:Install hpccsystems-platform-community_7.13.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | elaps | 
|---|---|---|---|---|---|---|
|unittest |146 |146 |0 |0 |0 |52 sec |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |1 sec |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |5 sec |



Regression test result:

| phase | total | pass | fail | elaps | 
|---|---|---|---|---|
|setup (hthor) |9 |9 |0 |21 sec  (00:00:21) |
|setup (thor) |9 |9 |0 |40 sec  (00:00:40) |
|setup (roxie) |19 |19 |0 |21 sec  (00:00:21) |
|test (hthor) |954 |954 |0 |664 sec  (00:11:04) |
|test (thor) |865 |865 |0 |1085 sec  (00:18:05) |
|test (roxie) |1027 |1027 |0 |724 sec  (00:12:04) |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|    7 sec (00:00:07) |  600 sec (00:10:00) |  106 sec (00:01:46) |   21 sec (00:00:21) |   16 sec (00:00:16) | 1392 sec (00:23:12) |   18 sec (00:00:18) | 2160 sec (00:36:00) |

ghalliday commented:
@jpmcmu see Richard's PR https://github.com/hpcc-systems/HPCC-Platform/pull/15785
